proposals/nnnn-trailing-comma-lists.md

```diff
@@ -0,0 +1,294 @@
++  b: Int,
+)
+```
+
+> [!NOTE]  
+> A similar proposal was [rejected](https://forums.swift.org/t/rejected-se-0084-allow-trailing-commas-in-parameter-lists-and-tuples/2777) back in 2016 for Swift 3. It's been 8 years since that, the swift language has evolved a lot, some changes highlighted above as motivation, and the code style that "puts the terminating right parenthesis on a line following the arguments to that call" has been widely adopted by community, swift standard library codebase, swift-format, docc documentation and Xcode. Therefore, not encourage or endorse this code style doesn't hold true anymore nor is a reason for rejection.
+
+## Proposed solution
+
+This proposal adds support for trailing comma for comma-separated lists, but only when there's a clear terminator. A enum case label list won't support, for example:
+
+```swift
+enum E {
+  case a, b, c, // ❌ Expected identifier after comma in enum 'case' declaration
+}
+```
+
+Trailing comma will be allowed in the following comma-separated lists:
+
+- Tuples and tuple patterns 
+
+```swift
+(1, 2,)
+let block: (Int, Int,) -> Void = { (a, b,) in  }
+let (a, b,) = (1, 2,)
+for (a, b,) in zip(s1, s2) { }
+```
+
+**Trailing comma will be allowed in single-element lists but not in zero-element lists.**
+
+```swift
+(1,) // OK
+(,) // ❌ expected value in tuple
+```
+
+- Initializers, functions and associated values arguments and parameters
+
+```swift
+
+m[x, y,]
+
+func foo(a: Int, b: Int,) { }
+
+struct S {
+    init(a: Int, b: Int,) { }
+}
+
+enum E {
+    case foo(a: Int, b: Int,)
+}
+```
+- Subscripts
+
+```swift
+let value = m[x, y,]
+```
+
+- `KeyPath` subscripts 
+
+```swift
+let keyPath = \Foo.bar[x,y,]
+          
+f(\.[x,y,])
+```
+
+- Attributes arguments
+
+```swift
+@Foo(1, 2, 3,) struct S { }
+f(_: @foo(1, 2,) Int)
+```
+
+- Macro expansion arguments
+
+```swift
+#foo(1, 2,)
+
+struct S {
+    #foo(1, 2,)
+}
+```
+
+- `if`, `guard` and `while` conditions
+
+```swift
+if a, b, { }
+while a, b, { }
+guard a, b, else { }
+```
+
+- `switch` case labels
+
+```swift
+switch number {
+    case 1, 2,:
+        ...
+    default:
+        ..
+}
+```
+
+- Closure capture list
+
+```swift
+{ [a, b,] in }
+```
+
+- inheritance list
+
+```swift
+struct S: P1, P2, P3, { }
+```
+
+**Inheritance lists won't support trailing comma when there isn't a clear terminator, such as `associatedtype` in a protocol declaration:**
+```swift
+protocol Foo {
+  associatedtype T: P1, P2, // ❌ Expected type
+}
+```
+
+- Generic parameters
+
+```swift
+struct S<T1, T2, T3,> { }
+```
+
+- Generic `where` clause list
+
+```swift
+struct S<T1, T2, T3> where T1: P1, T2: P2, { }
+```
+
+**Generic `where` clause lists won't support trailing comma when there isn't a clear terminator, such as functions in protocol declaration:**
+```swift
+protocol Foo {
+  func f<T1, T2>(a: T1, b: T2) where T1: P1, T2: P2, // ❌ Expected type
+}
+```
+
+- Availability spec list
+
+```swift
+if #unavailable(iOS 15, watchOS 9,) { }
```

Comment:
Point of clarification: since `*` is (required to be?) at the end of these lists if used, can you give an example (maybe best in the "Detailed design" section) regarding whether `#available(iOS 9999, *,)` will be supported—in other words, a trailing comma after `*`?